File: crates/zoro-core/src/workspace.rs
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use crate::{library::Library, query, Entry};

pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

/// 多库工作区：搜索 / 浏览的默认作用域。
///
/// 条目全局身份 = `(库名, path, start)`；聚合查询时库名作为命名空间前缀。
#[derive(Debug, Clone, Default)]
pub struct Workspace {
    pub libraries: Vec<Library>,
}

impl Workspace {
    /// 由 `name=path` 声明列表构建工作区（冷启动优先，过期自动重建）。
    pub fn from_libs(specs: &[(String, PathBuf)]) -> Result<Self> {
        let mut seen = BTreeSet::new();
        for (name, _) in specs {
            if !seen.insert(name.clone()) {
                return Err(format!("duplicate library name: {name}").into());
            }
        }
        let mut libraries = Vec::with_capacity(specs.len());
        for (name, path) in specs {
            libraries.push(Library::open_cached(name, path)?);
        }
        Ok(Self { libraries })
    }

    /// 单库便捷构造（`ZORO_ROOT` 场景）。
    pub fn single(name: &str, root: &Path) -> Result<Self> {
        Self::from_libs(&[(name.to_string(), root.to_path_buf())])
    }

    pub fn library_names(&self) -> Vec<String> {
        self.libraries
            .iter()
            .map(|l| l.name.clone())
            .collect()
    }

    /// 对全部库执行 analyze。`force=true` 无条件重扫。
    pub fn analyze_all(&mut self, force: bool) -> Result<()> {
        for lib in &mut self.libraries {
            lib.analyze(force)?;
        }
        Ok(())
    }

    /// 聚合查询，结果按分数降序（跨库）。
    pub fn query(&self, q: &str) -> Vec<query::Candidate> {
        let mut out = Vec::new();
        for lib in &self.libraries {
            out.extend(lib.query(q));
        }
        out.sort_by(|a, b| b.score.cmp(&a.score));
        out
    }

    /// 按三元组定位库并现场截取条目原文。
    pub fn load_raw(&self, library: &str, path: &Path, start: usize) -> Result<String> {
        let lib = self
            .libraries
            .iter()
            .find(|l| l.name == library)
            .ok_or_else(|| format!("library not found: {library}"))?;
        let entry = Entry {
            library: library.to_string(),
            path: path.to_path_buf(),
            start,
            ..Default::default()
        };
        lib.load_raw(&entry)
    }
}

File: core/workspace.go
package core

import (
	"fmt"
	"sort"
)

// Workspace is the default scope for search/browse across libraries.
//
// A block's global identity is (Library, Path, Start).
type Workspace struct {
	Libraries []*Library
}

// BuildWorkspace opens all declared libraries (cold-start first, auto rebuild).
func BuildWorkspace(cfg *WorkspaceConfig) (*Workspace, error) {
	if cfg == nil {
		return nil, fmt.Errorf("nil workspace config")
	}
	seen := map[string]bool{}
	for _, spec := range cfg.Libraries {
		if seen[spec.Name] {
			return nil, fmt.Errorf("duplicate library name: %s", spec.Name)
		}
		seen[spec.Name] = true
	}
	ws := &Workspace{make([]*Library, 0, len(cfg.Libraries))}
	for _, spec := range cfg.Libraries {
		lib, err := OpenLibraryCachedWithConfigAndData(spec.Name, spec.Root, spec.Config, cfg.DataDir)
		if err != nil {
			return nil, err
		}
		ws.Libraries = append(ws.Libraries, lib)
	}
	return ws, nil
}

// LoadWorkspace reads a `zoro.toml` file and builds a Workspace.
func LoadWorkspace(path string) (*Workspace, error) {
	cfg, err := WorkspaceConfigFromPath(path)
	if err != nil {
		return nil, err
	}
	return BuildWorkspace(&cfg)
}

// NewWorkspace builds a Workspace from simple name/root specs (tests / programmatic use).
func NewWorkspace(specs []LibrarySpec) (*Workspace, error) {
	return BuildWorkspace(&WorkspaceConfig{Libraries: specs})
}

// LibraryNames returns the declared library names in order.
func (w *Workspace) LibraryNames() []string {
	names := make([]string, 0, len(w.Libraries))
	for _, l := range w.Libraries {
		names = append(names, l.Name)
	}
	return names
}

// AnalyzeAll runs analyze on every library; force=true unconditionally rescans.
// It stops at the first failure.
func (w *Workspace) AnalyzeAll(force bool) error {
	for _, lib := range w.Libraries {
		if _, err := lib.Analyze(force); err != nil {
			return err
		}
	}
	return nil
}

// RefreshAll syncs every library with its content root and returns the first
// failure. Unlike Query, it does not swallow errors — frontends use it to tell
// the user why an index is unusable (e.g. core.ErrStoreLocked).
func (w *Workspace) RefreshAll() error {
	for _, lib := range w.Libraries {
		if err := lib.Refresh(); err != nil {
			return fmt.Errorf("刷新库 %s: %w", lib.Name, err)
		}
	}
	return nil
}

// Query aggregates cross-library candidates ranked by score desc.
func (w *Workspace) Query(q string) []Candidate {
	var out []Candidate
	for _, lib := range w.Libraries {
		out = append(out, lib.Query(q)...)
	}
	sort.SliceStable(out, func(i, j int) bool {
		return out[i].Score > out[j].Score
	})
	return out
}

// LoadRaw locates a library by name and cuts a block's raw text by (Path, Start).
func (w *Workspace) LoadRaw(library, path string, start int) (string, error) {
	for _, lib := range w.Libraries {
		if lib.Name == library {
			return lib.LoadRaw(&Block{Library: library, Path: path, Start: start})
		}
	}
	return "", fmt.Errorf("library not found: %s", library)
}
